### PR TITLE
Fix v0.7 deprecations and drop v0.6 support

### PR DESCRIPTION
**Pull request by jonathanBieler**
I first tried to make a version compatible with v0.6 using Compat but there's a bit too many changes that aren't in Compat. It seems a lot of packages are just dropping v0.6. This shouldn't be an issue since Memoize seems quite stable (the current release will keep working on v0.6).

### CODE REVIEW

**Comment by jonathanBieler** [comment]
I think I need to update the travis script for 1.0, I'll check that later.

**Comment by Element-126** [comment]
Fixes #30.

**Comment by simonster** [comment]
I'm totally on board with dropping support for <0.7, but if we're going to do that, would you mind deleting the code to support older versions?

**Comment by Element-126** [comment]
BTW, I raised the version to `v0.3.0` since we are obviously breaking compatibility with earlier versions of Julia.

**Comment by Element-126** [comment]
Hmm... the build is failing. What did I mess up ?

**Comment by Element-126** [comment]
> ERROR: Cannot develop package with the same name or uuid as the project

Has anyone seen this error yet ? I will google it and see how to work around it.

**Comment by Element-126** [comment]
Ok, the build script needs to be updated to support `Project.toml`. I will send a PR to @jonathanBieler's fork.

**Comment by Element-126** [comment]
There seems to be a good example [here](https://github.com/JuliaParallel/MPI.jl/pull/210/commits/d78d460f379bc67e3a4665acaf61e627138595b4) of how to configure the CI on 0.7+. I will try to imitate it and see how it goes.

**Comment by Element-126** [comment]
Could you try with this patch ? Not sure if it will work, but this is the recommended way ([see here](https://docs.travis-ci.com/user/languages/julia/)).

```patch
diff --git a/.travis.yml b/.travis.yml
index 45611d5..da15316 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -9,6 +9,6 @@ notifications:
   email: false
 script:
   - if [[ -a .git/shallow ]]; then git fetch --unshallow; fi
-  - julia --check-bounds=yes -e 'using Pkg; Pkg.clone(pwd()); Pkg.build("Memoize"); Pkg.test("Memoize"; coverage=true)'
+  - julia --check-bounds=yes -e 'using Pkg; Pkg.build(); Pkg.test()'
 after_success:
   - julia -e 'using Pkg; cd(Pkg.dir("Memoize")); Pkg.add("Coverage"); using Coverage; Coveralls.submit(Coveralls.process_folder())'
```

PS: I find it simpler to post patches here rather than open a PR each time, but if GitHub has a feature for that, please let me know.

**Comment by Element-126** [comment]
BTW, it might be easier to revert the change from `REQUIRE` to `Project.toml` in a first time, and do it later once there is a clear migration path and other projects have made the jump.

So feel free to `git rebase -i` it out so that this PR can move on.

**Comment by jonathanBieler** [comment]
Seems to work now.

**Comment by jonathanBieler** [comment]
Ok, it works now.

**Comment by Element-126** [comment]
Good job @jonathanBieler! I guess this is ready to be merged now.

File: test/runtests.jl [hunk 1]
```diff
@@ -218,7 +218,7 @@ function outer()
     @test run == 3
 end
 outer()
-@test !isdefined(:inner)
+@test !@isdefined inner
 
 if VERSION >= v"0.5.0-dev+5235"
```
[the remark this answers is not in the document]
Comment (Element-126) [reply test/runtests.jl]: You are right. I don't know how I missed this one. The following patch should do the trick:
```patch
diff --git a/test/runtests.jl b/test/runtests.jl
index 32aca48..924602c 100644
--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -220,13 +220,11 @@ end
 outer()
 @test !@isdefined inner
 
-if VERSION >= v"0.5.0-dev+5235"
-    @memoize function typeinf(x)
-        x + 1
-    end
-    @inferred typeinf(1)
-    @inferred typeinf(1.0)
+@memoize function typeinf(x)
+    x + 1
 end
+@inferred typeinf(1)
+@inferred typeinf(1.0)
 
 println("The following method rewrite warnings are normal")
 finalized = false
```
I just checked that this was the last occurrence of `VERSION` in the package. There may still be some legacy code lying around, but I am not able to see it. Maybe `femtocleaner` would be helpful here.

**Comment by jonathanBieler** [comment]
Done.

**Comment by simonster** [comment]
Thanks!

**Pull request merged**